src/meta-srv/src/events.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use client::{Client, Database};
+use common_catalog::consts::{DEFAULT_CATALOG_NAME, DEFAULT_PRIVATE_SCHEMA_NAME};
+use common_error::ext::BoxedError;
+use common_event_recorder::error::{
+    InsertEventsSnafu, KvBackendSnafu, NoAvailableFrontendSnafu, Result,
+};
+use common_event_recorder::{build_row_inserts_request, insert_hints, Event, EventHandler};
+use common_grpc::channel_manager::ChannelManager;
+use common_meta::peer::PeerLookupServiceRef;
+use common_procedure::event::ProcedureEvent;
+use common_telemetry::debug;
+use snafu::{ensure, ResultExt};
+
+use crate::cluster::MetaPeerClientRef;
+use crate::events::region_migration_event::RegionMigrationEvent;
+use crate::lease::MetaPeerLookupService;
+
+pub mod region_migration_event;
+
+/// EventHandlerImpl is the default event handler implementation in metasrv. It sends the received events to the frontend instances.
+pub struct EventHandlerImpl {
+    peer_lookup_service: PeerLookupServiceRef,
+    channel_manager: ChannelManager,
+}
+
+impl EventHandlerImpl {
+    pub fn new(meta_peer_client: MetaPeerClientRef) -> Self {
+        Self {
+            peer_lookup_service: Arc::new(MetaPeerLookupService::new(meta_peer_client)),
+            channel_manager: ChannelManager::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl EventHandler for EventHandlerImpl {
+    async fn handle(&self, events: &[Box<dyn Event>]) -> Result<()> {
+        let region_migration_events: Box<Vec<Box<dyn Event>>> = Box::new(
+            events
+                .iter()
+                .filter_map(|event| {
+                    event
+                        .as_any()
+                        .downcast_ref::<ProcedureEvent>()
+                        .and_then(|procedure_event| {
+                            RegionMigrationEvent::new_from_procedure_event(procedure_event)
+                                .ok()
+                                .flatten()
+                                .map(|event| Box::new(event) as Box<dyn Event>)
+                        })
+                })
+                .collect(),
```

Comment:
Can we add a method to the `Procedure` trait to expose recordable metadata?  And we can build the `Event` directly in the recorder without having to convert it to another `Event` implementation in the handler.